### PR TITLE
Bug 1169340 - Perf Compare should validate revisions before loading

### PR DESCRIPTION
**Pull request by MikeLing**
[Review on Reviewable](https://reviewable.io/reviews/mozilla/treeherder/588)

### CODE REVIEW

**Comment by jmaher** [comment]
doing a drive by here- this looks good!  I am happy to see this fixed:)  How does this work for the end user?  Is there a screenshot of how this looks?  When we have the error, this is when they hit the compare button, not when the input happens.  I am not sure if we need it to be dynamic.

How will this work if the url is entered with invalid values?  Maybe that could be addressed in another PR or bug- maybe this solves it as well or could easily do so.

**Comment by MikeLing** [comment]
Hi jmaher,

it's pity that I can't paste screenshot for now, because it's midnight in here and I'm afk already. But I can paste them tomorrow: )

>How will this work if the url is entered with >invalid values?

Do you mean put revision number in URL directly?    Hm, I'm not sure about it actually. Maybe I could try it and reply you after then.

Thank you

**Comment by MikeLing** [comment]
Hi jmaher, @jmaher 

I will paste some screen blow, please fell free to tell me your opinion, thank you! :)

When I type error revision number but haven't type compare button.
before i type compare

Let's press it ;)
with both error

It can also deal with single error
one error

I also try put revision number directly, and it will jump into this page :)
put it in url d

**Comment by jmaher** [comment]
These look great!  I think a followup bug should be to make invalid URL's
end up on comparechooser with the correct error message :)  Out of scope
for this.

File: ui/js/compareperf.js [hunk 1]
```diff
@@ -5,19 +5,35 @@
...
+               if($scope.originalRevisionError == undefined && $scope.newRevisionError == undefined) {
+                $state.go('compare', {
+                originalProject: $scope.originalProject.name,
+                originalRevision: $scope.originalRevision,
+                newProject: $scope.newProject.name,
+                newRevision: $scope.newRevision });        
+                }
+        }, function(error) {
+                $scope.newRevisionError = error;
+        });
```
Comment (tojonmz) [inline ui/js/compareperf.js]: Curious @MikeLing did this get run through a js linter? To my (untrained :)) eye, there appears to be some indentation issues above.

File: ui/partials/perf/comparechooserctrl.html [hunk 2]
```diff
@@ -16,7 +20,11 @@
           <label for="new-project-selector">Project</label>
           <select id="new-project-selector" class="form-control" ng-model="newProject" ng-options="project.name for project in projects"/>
           <label for="new-revision-input">Revision</label>
-          <input id="new-revision-input" maxlength="12" class="form-control" ng-model="newRevision">
+            <div class="form-group has-error">
+            <input id="new-revision-input" maxlength="12" class="form-control" ng-model="newRevision">
+            <br/>
```
Comment (tojonmz) [inline ui/partials/perf/comparechooserctrl.html]: If these two `input` and `br` lines are child elements of the parent div, I believe they should be indented 2 chars.

Comment (MikeLing) [inline ui/partials/perf/comparechooserctrl.html]: ah, sorry about that! I always have code indent problem:(  I will fix this tomorrow. Can you tell me some tools can check out my code style like flask8 for Python? 

By the way, can we talk on irc about the other bug I work on? I still confuse about ng-class and I really want to know how to solve that. Thank you:)

**Comment by tojonmz** [comment]
> Let's press it ;)

Maybe outside the scope of this bug, but it seems a bit unnecessary to say "No results for revision *7atf afaftda6ffa6df* on *mozilla-central*" when both values are directly above the error?

For simplicity I would think
* *"Invalid revision"*
* *"No results"*
* *"Revision not found"*

or similar, as an error might be sufficient here.

File: ui/partials/perf/comparechooserctrl.html (hunk 2 above)
Comment (tojonmz) [inline ui/partials/perf/comparechooserctrl.html]: > ah, sorry about that! I always have code indent problem:( I will fix this tomorrow. Can you tell me some tools can check out my code style like flask8 for Python?

http://old.jslint.com
http://www.jslint.com (new, beta.. still a bit strange)

Comment (tojonmz) [inline ui/partials/perf/comparechooserctrl.html]: Oh, and on that line 25, I'm not sure about closing that `<br/>`, I defer to @wlach. I see in the google style guide for HTML5, they indicate void tags should not receive one.

*"Although fine with HTML, do not close void elements, i.e. write \<br>, not \<br /*>."*
http://google-styleguide.googlecode.com/svn/trunk/htmlcssguide.xml#HTML_Style_Rules > Document Type

**Comment by tojonmz** [comment]
@MikeLing @jmaher I had a chance to also just test this branch now (post-merge). Fwiw a couple of other things I think could be improved :) :

The default landing appearance surrounds the inputs with red warning borders, when I haven't done anything yet that is 'bad'. Imo it should only surround with red after I submit a 12 character SHA which is not available after the submit?

(default landing appearance)
defaultappearance

Bit of an edge case below, but the form doesn't appear to do alphanumeric checking for the SHA format. For example I can enter these 12 character values, but afaik they are bogus SHA's with the trailing special character `*`. The 'Compare Revisions' button should not turn dark blue unless the 12 characters are either letters or numbers.

nodatavalidation
And lastly, if I enter some bad SHA's and hit submit, and then enter some new SHA's, the warnings persist until I click Compare Revisions again. Note I have entered new, valid SHA's below that differ from the stale warnings.

stalewarningmessages

Takeaway: I wonder if we even need the lengthy text dialogue in red. Wouldn't just using red input field surrounds, a tiny red "(!)" above the top right corner of each input field and a tooltip message "Revision not found" be sufficient?

**Comment by MikeLing** [comment]
Hi @tojonmz ,
I do have a version commit before that can show blue when you put right and show red when your revision is wrong.

screenshot from 2015-06-08 23 18 44
screenshot from 2015-06-08 23 16 00

But, the method I use is put a "ng-class" like "ng-class="{'has-error':originalRevisionError}"", but Will tells me he is not sure if it's appropriate method and I should follow the bootstrap document. So I just change it back.

**Comment by tojonmz** [comment]
Fwiw, this is my take on what it should look like, per my ealier comment and screen grabs:

> Takeaway: I wonder if we even need the lengthy text dialogue in red. Wouldn't just using red input field surrounds, a tiny red "(!)" above the top right corner of each input field and a tooltip message on each input field "Revision not found" be sufficient?

**Pull request merged**